pyo3-derive-backend/src/proto_method.rs

```diff
@@ -1,70 +1,59 @@
 // Copyright (c) 2017-present PyO3 Project and Contributors
-use crate::utils::print_err;
 use proc_macro2::{Span, TokenStream};
 use quote::{quote, ToTokens};
 use syn::Token;
 
 // TODO:
 //   Add lifetime support for args with Rptr
 #[derive(Debug)]
-pub enum MethodProto {
-    Free {
-        name: &'static str,
-        proto: &'static str,
-    },
-    Unary {
-        name: &'static str,
-        proto: &'static str,
-    },
-    UnaryS {
-        name: &'static str,
-        arg: &'static str,
-        proto: &'static str,
-    },
-    Binary {
-        name: &'static str,
-        arg: &'static str,
-        proto: &'static str,
-    },
-    BinaryS {
-        name: &'static str,
-        arg1: &'static str,
-        arg2: &'static str,
-        proto: &'static str,
-    },
-    Ternary {
-        name: &'static str,
-        arg1: &'static str,
-        arg2: &'static str,
-        proto: &'static str,
-    },
-    TernaryS {
-        name: &'static str,
-        arg1: &'static str,
-        arg2: &'static str,
-        arg3: &'static str,
-        proto: &'static str,
-    },
-    Quaternary {
-        name: &'static str,
-        arg1: &'static str,
-        arg2: &'static str,
-        arg3: &'static str,
-        proto: &'static str,
-    },
+pub struct MethodProto {
+    pub name: &'static str,
+    pub args: &'static [&'static str],
+    pub proto: &'static str,
+    pub with_self: bool,
+    pub with_result: bool,
 }
 
-impl MethodProto {
-    pub fn name(&self) -> &str {
-        match *self {
-            MethodProto::Free { ref name, .. } => name,
-            MethodProto::Unary { ref name, .. } => name,
-            MethodProto::UnaryS { ref name, .. } => name,
-            MethodProto::Binary { ref name, .. } => name,
-            MethodProto::BinaryS { ref name, .. } => name,
-            MethodProto::Ternary { ref name, .. } => name,
-            MethodProto::TernaryS { ref name, .. } => name,
-            MethodProto::Quaternary { ref name, .. } => name,
+pub struct MethodProtoBuilder {
+    pub name: &'static str,
+    pub proto: &'static str,
+    pub args: &'static [&'static str],
+    pub with_self: bool,
+    pub with_result: bool,
+}
+
+impl MethodProtoBuilder {
+    pub const fn new(name: &'static str, proto: &'static str) -> Self {
+        MethodProtoBuilder {
+            name,
+            proto,
+            args: &[],
+            with_self: false,
+            with_result: true,
+        }
+    }
+    pub const fn args(self, args: &'static [&'static str]) -> MethodProtoBuilder {
+        let mut with_args = self;
+        with_args.args = args;
+        with_args
+    }
```

Comment:
I think you can probably take `mut self` in each of these builder methods, e.g.

```suggestion
    pub const fn args(mut self, args: &'static [&'static str]) -> MethodProtoBuilder {
        self.args = args;
        self
    }
```

Reply:
You're right, I didn't think `mut self` was possible but the code is much nicer this way.